src/backends/torch/torchinputconns.cc

```diff
@@ -508,67 +540,69 @@ namespace dd
+                  {
+                    add_image_file<int>(_test_dataset, lfile.first,
+                                        lfile.second);
+                  }
 
-                _logger->info(
-                    "data split test size={} / remaining data size={}",
-                    test_lfiles.size(), lfiles.size());
+                // Write corresp file
+                std::ofstream correspf(_model_repo + "/" + _correspname,
+                                       std::ios::binary);
+                auto hit = hcorresp.begin();
+                while (hit != hcorresp.end())
+                  {
+                    correspf << (*hit).first << " " << (*hit).second
+                             << std::endl;
+                    ++hit;
+                  }
+                correspf.close();
               }
-
-            // Read data
-            for (const std::pair<std::string, int> &lfile : lfiles)
+            else // file exists, expects a list of files and targets, for
+                 // regression & multi-class
               {
-                add_image_file(_dataset, lfile.first, lfile.second);
-              }
+                std::vector<std::pair<std::string, std::vector<double>>>
+                    lfiles; // labeled files
+                std::vector<std::pair<std::string, std::vector<double>>>
+                    test_lfiles;
+                read_image_list(lfiles, _uris.at(0));
+                if (_uris.size() > 1)
+                  read_image_list(test_lfiles, _uris.at(1));
 
-            for (const std::pair<std::string, int> &lfile : test_lfiles)
-              {
-                add_image_file(_test_dataset, lfile.first, lfile.second);
-              }
+                bool has_test_data = test_lfiles.size() != 0;
+                if (_test_split > 0.0 && !has_test_data)
+                  {
+                    split_dataset<std::vector<double>>(lfiles, test_lfiles);
+                  }
 
-            // Write corresp file
-            std::ofstream correspf(_model_repo + "/" + _correspname,
-                                   std::ios::binary);
-            auto hit = hcorresp.begin();
-            while (hit != hcorresp.end())
-              {
-                correspf << (*hit).first << " " << (*hit).second << std::endl;
-                ++hit;
+                // Read data
+                for (const std::pair<std::string, std::vector<double>> &lfile :
+                     lfiles)
+                  {
+                    add_image_file<std::vector<double>>(_dataset, lfile.first,
+                                                        lfile.second);
+                  }
+
+                for (const std::pair<std::string, std::vector<double>> &lfile :
+                     test_lfiles)
+                  {
+                    add_image_file<std::vector<double>>(
+                        _test_dataset, lfile.first, lfile.second);
+                  }
```

Comment:
I think it would be great to have dataset file list building and test split in ImgInputConnector, while keeping conversion to tensor and saving to database in the Torch connector. I see two ways of doing that: either we pass a flag to img input connector so that it only reads uris and let the connector read the image files and put them in the database, or ImgInputConnector preprocesses images and call a hook to put the images in db instead of saving them to `ImgInputConnector::_images`. Or something else ?

This can be done in a future PR as well.

Reply:
Sure this can be discussed outside this PR. Mainly caffe and torch backends allow training. So refactoring would remove local code which is good.